### ISSUE TITLE
Doubts about bilinear interpolation in presence of ROI or nodata

### ISSUE THREAD

**Issue by aaime** [geosolutions-it/jai-ext#203]
The current code seems to be basically an adaptation of the path with all pixels available, I think as coded is occasionally causing darker pixels to show up in the output on the borders of ROIs.

https://github.com/geosolutions-it/jai-ext/blob/14c227b2fcd1e841f10a6e74f8cfd667cb6fc944/jt-utilities/src/main/java/it/geosolutions/jaiext/interpolators/InterpolationBilinear.java#L930-L980

In particular, the code does two horizonal interpolations, and then performs a vertical one:
* If both of the two upper pixels are outside or ROI, the upper hor interpolation is set to zero, and then this zero is used in the vertical interpolation. Instead, I believe the yFrac should be checked, if the interpolated points is inside the top area, nodata should be returned, if in the lower area, the interpolation of the lower area should be used instead
* Same happens if both lowers are missing
* No logic covers the case of both left or both right pixels are outside of the ROI or nodata, but imho the logic should be similar to the ones above for top and bottom
* In case just one of the top pixels is valid, the code is still multiplying the value by the associated x fraction or complementary x fraction... but this just lowers the value, darkening it. Imho it should be kept at is instead. Same goes for the bottom ones

**Comment by aaime** [comment]
Actually, some of the logic I was describing is implemented, but only for the non integer case! (the integer case is the first one found in both the scale op and the stand alone bilinear interpolator!

**Comment by aaime** [comment]
Btw, as a visual aid, I believe the residual issues are the cause of the darker 1px shadows at the border of this picture (there is snow there, should be white):

![image](https://user-images.githubusercontent.com/325433/44530144-0e786a80-a6ee-11e8-8810-0496d6a598fb.png)

**Comment by aaime** [comment]
A bit more information. I've saved the images from the rendered image browser and replicated the chain with all the same parameters in a stand alone test, and the shadows do not happen.
I've also took the same image as above, but instead of picking it out of a single image mosaic, I published it as geotiff, and the issue also does not happen.
So it seems it's something specific to the usage in the context of image mosaic, but mind, it's still happening at the level of "scale after read", here is a screenshot:

![selezione_358](https://user-images.githubusercontent.com/325433/44576527-12a89480-a78f-11e8-918e-c6dce1da3a58.png)

and the same reading from the RIB "save image" dumps and running a short chain out of java code, with all same parameters, same operations, same hints... 

![selezione_360](https://user-images.githubusercontent.com/325433/44577020-4cc66600-a790-11e8-96aa-fae60685579a.png)

Checked, the ROI look exactly the same, the params appear to be the same, the images have same origin and size, and yet the interpolation at the borders of the ROI seems to work differently. Also checked they are getting in exactly the same code base, "caseB", the part not using ROI accessor.
